x-pack/plugins/canvas/public/state/actions/elements.js

```diff
@@ -213,12 +214,48 @@ export const duplicateElement = createThunk(
   }
 );
 
+export const rawDuplicateElement = createThunk(
+  'rawDuplicateElement',
+  ({ dispatch, type }, element, pageId, root) => {
+    const newElement = cloneDeep(element);
+    // move the root element so users can see that it was added
+    newElement.position.top = newElement.position.top + 10;
+    newElement.position.left = newElement.position.left + 10;
+    const _rawDuplicateElement = createAction(type);
+    dispatch(_rawDuplicateElement({ pageId, element: newElement }));
+
+    // refresh all elements if there's a filter, otherwise just render the new element
+    if (element.filter) {
+      dispatch(fetchAllRenderables());
+    } else {
+      dispatch(fetchRenderable(newElement));
+    }
+
+    // select the new element
+    if (root) {
+      window.setTimeout(() => dispatch(selectElement(newElement.id)));
+    }
+  }
+);
+
 export const removeElements = createThunk(
   'removeElements',
-  ({ dispatch, getState }, elementIds, pageId) => {
+  ({ dispatch, getState }, rootElementIds, pageId) => {
+    const state = getState();
+
+    // todo consider doing the group membership collation in aeroelastic, or the Redux reducer, when adding templates
```

Comment:
Consider matching a TODO with an issue number, even if it's a single issue that encompasses all of the follow-ups.  This makes it easier to `grep` for all of the changes you want to make, (since searching for `todo` would probably yield far too many results)